google_permissions/examples/basic_with_addon/main.tf

```diff
@@ -0,0 +1,14 @@
+module "permissions" {
+  source = "../../../google_permissions"
+  # it is assumed that you loaded and have available a local.project
+  google_folder_id  = local.project.folder.id
+  google_prod_id    = local.project["prod"].id
+  google_nonprod_id = local.project["nonprod"].id
+  admin_members     = ["workgroup:my-project/admins"]
+  developer_members = ["workgroup:my-project/developers"]
+  other_permissions = [
```

Comment:
```suggestion
  nonprod_roles = [
  prod_roles = [
```

Reply:
@jasonthomas  - one thing first - I created two distinct sets of allowed roles in case we want specific targeting different environments. Not sure if necessary, but seemed a good idea?

Second, did some homework on the solution. I'll go with option 3, but wanted to run it by you first.

```hcl
# see https://github.com/hashicorp/terraform/issues/25609#issuecomment-1472119672

# Three ways to do the validation of a list of roles against an allowed list


locals {
  # This is a list of all the permissions that we allow to be added to the project.
  # If you want to add a new permission, add it to this list.
  prod_allowed_roles = [
    "roles/bigquery.jobUser",
    "roles/automl.editor",
    "roles/cloudtranslate.editor",
    "roles/storage.objectAdmin",
    "roles/translationhub.admin",
    "roles/storage.admin",
    "roles/cloudsql.admin"
  ]

  nonprod_allowed_roles = [
    "roles/bigquery.jobUser",
    "roles/editor",
    "roles/storage.admin",
    "roles/cloudsql.admin"
  ]
}

variable "prod_roles" {
  description = "List of roles to apply to the prod project."
  type        = list(string)
  default     = []

  # WAY #1 - don't use locals -- have it all here, BUT then have to maintain list of roles in this block
  validation {
    condition = alltrue([
      for x in var.prod_roles : contains([
        "roles/bigquery.jobUser",
        "roles/automl.editor",
        "roles/cloudtranslate.editor",
        "roles/storage.objectAdmin",
        "roles/translationhub.admin",
        "roles/storage.admin",
        "roles/cloudsql.admin"
      ], x)
    ])
    error_message = "You have specified an invalid prod role."
  }
}

# WAY #2 - locals with dummy variable. throws kinda sketchy error
locals {
  validate_prod_roles = alltrue([
    for x in var.prod_roles : contains(local.prod_allowed_roles, x)
  ]) ? true: tobool("You have specified an invalid prod role.")
}

variable "nonprod_roles" {
  description = "List of roles to apply to the nonprod project."
  type        = list(string)
  default     = []
  }

}

# WAY #3 - recommended in the bug above, same kind of sketchy error though
output "validate_nonprod_roles" {
  value = null

  precondition {
    condition = alltrue([
      for x in var.nonprod_roles : contains(local.nonprod_allowed_roles, x)
    ])
    error_message = "You have specified an invalid nonprod role."
  }
}

##################################################  
#
# Here's the way the error looks for the third way. Also similar in second. You get the error msg though.
# 
#Plan: 6 to add, 0 to change, 5 to destroy.
#╷
#│ Error: Module output value precondition failed
#│ 
#│   on ../../../../../mozilla/terraform-modules/google_permissions/variables.tf line 82, in output "validate_nonprod_roles":
#│   82:     condition = alltrue([
#│   83:       for x in var.nonprod_roles : contains(local.nonprod_allowed_roles, x)
#│   84:     ])
#│     ├────────────────
#│     │ local.nonprod_allowed_roles is tuple with 4 elements
#│     │ var.nonprod_roles is list of string with 2 elements
#│ 
#│ You have specified an invalid nonprod role.
```

Reply:
re: nonprod_roles + prod_roles question above -- there is also folder level roles which is not captured

Reply:
Using precondition seems like a reasonable approach, but this could also be used with a resource definition instead of the output like:

```
# we create one resources for all nonprod project role bindings, and enable/disable on admin_only variable
# what's nice about this approach is the precondition is not evaluated when admin_only is true. 
resource "google_project_iam_binding" "nonprod_roles" {
  for_each = toset([for role in var.nonprod_roles : x if !var.admin_only])
  project  = var.google_nonprod_id
  role     = each.value
  members  = module.developers_workgroup.members

  lifecycle {
    precondition {
      condition = alltrue([
        for x in var.nonprod_roles : contains(local.nonprod_allowed_roles, x)
      ])
      error_message = "You have specified an invalid nonprod role."
    }
  }
}
```

Reply:
But also noisy because precondition will fail for each resource attempting to be created.